daemons/ipa-slapi-plugins/ipa-pwd-extop/prepost.c

```diff
@@ -379,14 +385,19 @@ static int ipapwd_pre_add(Slapi_PBlock *pb)
     pwdop->pwdata.timeNow = time(NULL);
     pwdop->pwdata.target = e;
 
-    ret = ipapwd_CheckPolicy(&pwdop->pwdata);
-    /* For accounts created by cn=Directory Manager or a passsync
-     * managers, ignore result of a policy check */
-    if ((pwdop->pwdata.changetype != IPA_CHANGETYPE_DSMGR) &&
-        (ret != 0) ) {
-        errMesg = ipapwd_error2string(ret);
-        rc = LDAP_CONSTRAINT_VIOLATION;
-        goto done;
+    /* If the entry represents an app password, no password policy checks
+     * are performed
+     */
+    if (!slapi_sdn_isgrandparent(krbcfg->apppws_sdn, sdn)) {
+        ret = ipapwd_CheckPolicy(&pwdop->pwdata);
+        /* For accounts created by cn=Directory Manager or a passsync
+         * managers, ignore result of a policy check */
+        if ((pwdop->pwdata.changetype != IPA_CHANGETYPE_DSMGR) &&
+            (ret != 0)) {
+            errMesg = ipapwd_error2string(ret);
+            rc = LDAP_CONSTRAINT_VIOLATION;
+            goto done;
+        }
     }
```

Comment:
As for enforcement of app password generation on the LDAP level, it seems to me that it will be best to add an `else` branch here, where I will generate the app password and ensure it will then be:
1. stored into the LDAP database - this will be accomplished by simply putting the generated password into the entry to be stored (in the `e` variable).
2. somehow retrieved in the POST callback of the `apppw-add` method for displaying to the user - and here I do not know how to do this, as that is in a completely different file written in Python, not C (`ipaserver/plugins/apppw.py`). Currently, it is done in that file, but that is easy, as the app password is generated in the PRE callback, and then just saved and retrieved for display using `setattr(context, 'randompassword', ...)` and `getattr(context, 'randompassword')`.